010-server-side-rendering.md

```diff
@@ -0,0 +1,205 @@
+
+Gatsby is only able to statically render HTML, and so _everything_ we want in the server's response needs to be known at build time – requiring us to somewhat tediously tally up all of our content every time we build the site. I've heard this can get very slow (in some cases, hour-long builds!), although [Gatsby Build seems to help](https://www.gatsbyjs.org/blog/2020-01-27-announcing-gatsby-builds-and-reports/#builds-making-deploys-as-fast-as-our-websites)
+
+Ideally, we could build each page on-demand as it changes so that content changes would remain fast. We'd still need to run a full build for app changes, though, and incremental builds for Gatsby [are not yet shipped](https://www.gatsbyjs.org/blog/2020-01-27-announcing-gatsby-builds-and-reports/#are-builds-the-same-thing-as-incremental-builds).
+
+
+#### On-Demand Rendering
+
+Our website also often veers into "application" territory, since many of our pages have stateful interfaces (for example, the campaign page changes based on whether you're signed up or not). Even some otherwise simple "content" pages have dynamic elements, like how our [referral page](https://www.dosomething.org/us/join?user_id=5543dfd6469c64ec7d8b46b3&campaign_id=9037) is customized with the referrer's first name. We could hardly pre-render that page for every single user that might possibly refer someone!
+
+For dynamic content like this, the easiest solution is to render pages or components on-demand: each time a user visits a page, the server renders the React application to HTML and sends that as the response. The user's browser can display that markup as soon as it's downloaded & then render any interaction or navigation from there on the client-side.
+
+At first glance, we're just moving the same work from the client to the (admittedly, often faster) server. However, we've got a trick up our sleeve – caching! Once we've rendered a page once, we can store it in a public cache (on [Fastly](https://www.fastly.com) or [Zeit's CDN](https://zeit.co/smart-cdn)), making subsequent visits to that page "render" instantaneously.
+
+#### Static Rendering & On-Demand Rendering
+
+Next.js supports [opt-in static generation](https://nextjs.org/docs/basic-features/data-fetching#getstaticprops-static-generation) alongside on-demand rendered routes. This means that pages that are completely static (say, an article page) can be rendered to HTML at build-time while other pages continue to be rendered on-demand. Unfortunately, this [doesn't work with Apollo's server-side rendering](#) & so we're stuck with a choice between static generation & pre-rendering colocated queries.
```

Comment:
👾 It looks like the `doesn't work with Apollo's server-side rendering` link is missing.

Reply:
I didn't really understand the Apollo server-side rendering issue -- does it not work with on-demand rendering? So we'd have to go with static rendering?

Reply:
Gah yeah! I meant to write up more details on that in my proof-of-concept PR and link here, but totally forgot! I'll make myself a note to do that later today.

The TLDR is that when we use Apollo's ability to "render the tree" and fetch up any queries within a view, we aren't able to "optimize" to a statically-rendered page; so we have to choose an "either or" for those two approaches (for now, at least).

Reply:
I think I'm still confused about this Apollo piece - it can't do static generation but it does generally work with Next.js?

Reply:
> This means that pages that are completely static (say, an article page) can be rendered to HTML at build-time

How does this type of rendering work with content changes, is it the on-demand rendering you referenced above?

Reply:
We'd have to [re-trigger a build](https://www.contentful.com/developers/docs/tutorials/general/automate-site-builds-with-webhooks/) whenever content changes in order to rebuild all these static HTML pages. (We could perhaps consider using the [`fallback`](https://nextjs.org/docs/basic-features/data-fetching#fallback-false) option for `getStaticPaths` to get the "best of both worlds" and lazily re-render those static pages once they get a request if these builds get slow...)

There's also [Preview Mode](https://nextjs.org/docs/advanced-features/preview-mode), which would let editors bypass static generation (for viewing things on [preview.dosomething.org](https://preview.dosomething.org) before they're actually published).

Reply:
And finally, added that [promised write-up](https://github.com/DoSomething/phoenix-next/pull/1970#issuecomment-623726651) on the issues I ran into with `getStaticProps` (for static generation) and Apollo's server-side rendering support in 323f31f.